[Navigation] navigation-history-entry/entries-after-blob-navigation.html is failing

cdumez opened the pull request:
#### f183cbdda7131eaf27997de4a3e4e0905cc2203b
<pre>
[Navigation] navigation-history-entry/entries-after-blob-navigation.html is failing
<a href="https://bugs.webkit.org/show_bug.cgi?id=282609">https://bugs.webkit.org/show_bug.cgi?id=282609</a>
<a href="https://rdar.apple.com/139725684">rdar://139725684</a>

Reviewed by Rob Buis.

Resync entries-after-blob-navigation.html from upstream to get the fix so that the
blob&apos;s content type is provided. This allows the test to run though the test would
still fail. The test was navigating the iframe in the top frame&apos;s load event handler
and expecting a new HistoryItem to get created for the navigation. However, we had
non-standard logic in `NavigationScheduler::mustLockBackForwardList()` that would
lock the back/forward list if any of the iframe&apos;s ancestors is still loading. Since
this behavior doesn&apos;t match the HTML specification [1][2] or Blink, remove the logic
in mustLockBackForwardList().

The change above caused a different code path to get taken for
navigation-api/navigation-history-entry/entries-after-srcdoc-navigation.html, which
caused it to fail because it would fail to create a new HistoryItem for the srcdoc
navigation. After some investigation, I found that both the back/forward list and the
history list were unlocked. However. we would prevent the creation of a HistoryItem
for a srcdoc navigation because the substitute data created for the navigation had the
`SubstituteData::SessionHistoryVisibility::Hidden` parameter. As per the tests, we
expect srcdoc navigations to be visible in the History list so I am setting this
parameter to &quot;Visible&quot; instead. This got the test passing.

[1] <a href="https://html.spec.whatwg.org/#navigate-an-iframe-or-frame">https://html.spec.whatwg.org/#navigate-an-iframe-or-frame</a> (step 2)
[2] <a href="https://html.spec.whatwg.org/#completely-finish-loading">https://html.spec.whatwg.org/#completely-finish-loading</a>

* LayoutTests/contentfiltering/block-after-add-data-expected.txt:
* LayoutTests/contentfiltering/block-after-add-data-then-allow-unblock-expected.txt:
* LayoutTests/contentfiltering/block-after-add-data-then-deny-unblock-expected.txt:
* LayoutTests/contentfiltering/block-after-finished-adding-data-expected.txt:
* LayoutTests/contentfiltering/block-after-finished-adding-data-then-allow-unblock-expected.txt:
* LayoutTests/contentfiltering/block-after-finished-adding-data-then-deny-unblock-expected.txt:
* LayoutTests/contentfiltering/block-after-response-expected.txt:
* LayoutTests/contentfiltering/block-after-response-then-allow-unblock-expected.txt:
* LayoutTests/contentfiltering/block-after-response-then-deny-unblock-expected.txt:
* LayoutTests/contentfiltering/block-after-will-send-request-expected.txt:
* LayoutTests/contentfiltering/block-after-will-send-request-then-allow-unblock-expected.txt:
* LayoutTests/contentfiltering/block-after-will-send-request-then-deny-unblock-expected.txt:
* LayoutTests/http/tests/navigation/javascriptlink-subframeload-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/history-traversal/srcdoc/consecutive-srcdoc-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/history-traversal/srcdoc/srcdoc-history-entries-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-history-entry/entries-after-blob-navigation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-history-entry/entries-after-blob-navigation.html:
* LayoutTests/platform/gtk/http/tests/navigation/javascriptlink-frames-expected.txt:
* LayoutTests/platform/mac-ventura/http/tests/navigation/javascriptlink-frames-expected.txt:
* LayoutTests/platform/mac-wk1/contentfiltering/block-after-will-send-request-expected.txt: Copied from LayoutTests/contentfiltering/block-after-will-send-request-expected.txt.
* LayoutTests/platform/mac-wk1/contentfiltering/block-after-will-send-request-then-allow-unblock-expected.txt: Copied from LayoutTests/contentfiltering/block-after-will-send-request-then-allow-unblock-expected.txt.
* LayoutTests/platform/mac-wk1/contentfiltering/block-after-will-send-request-then-deny-unblock-expected.txt: Copied from LayoutTests/contentfiltering/block-after-will-send-request-then-deny-unblock-expected.txt.
* LayoutTests/platform/mac/http/tests/navigation/javascriptlink-frames-expected.txt:
* LayoutTests/platform/wpe/http/tests/navigation/javascriptlink-frames-expected.txt:
* Source/WebCore/html/HTMLFrameElementBase.cpp:
(WebCore::HTMLFrameElementBase::attributeChanged):
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::defaultSubstituteDataForURL):
* Source/WebCore/loader/NavigationScheduler.cpp:
(WebCore::NavigationScheduler::mustLockBackForwardList):

Canonical link: <a href="https://commits.webkit.org/287549@main">https://commits.webkit.org/287549@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5c53b6d78f726e049b1286400fcfdae4d7adfb25

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/80007 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/59004 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/33405 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/84522 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/30980 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/82115 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/68068 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/7300 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/62539 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/20364 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/83075 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/52602 "Found 1 new test failure: media/audioSession/getUserMedia.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/72863 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/42848 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/49940 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/27016 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/29443 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/71060 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/27508 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/85954 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/7223 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/5086 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/70811 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/7400 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/68704 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/70053 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/14050 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/12985 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12383 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/7186 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/12718 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/7032 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/10546 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/8836 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rwlbuis commented on LayoutTests/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/history-traversal/srcdoc/consecutive-srcdoc-expected.txt:
```diff
@@ -1,4 +1,6 @@
+Harness Error (TIMEOUT), message = null
+
+PASS changing srcdoc does a replace navigation since the URL is still about:srcdoc
+TIMEOUT changing srcdoc to about:srcdoc#yo then another srcdoc does two push navigations and we can navigate back Test timed out
```
Would be nice to fix the timeout I guess, but these seem edge cases.

cdumez replied:
I should have specified in the commit log but these tests are basically proceeding further now (and failing later). This is actually a progression though I agree it'd be nice if it didn't time out.

rwlbuis reviewed (approved):
LGTM.

cdumez committed:
[Navigation] navigation-history-entry/entries-after-blob-navigation.html is failing
https://bugs.webkit.org/show_bug.cgi?id=282609
rdar://139725684

Reviewed by Rob Buis.

Resync entries-after-blob-navigation.html from upstream to get the fix so that the
blob's content type is provided. This allows the test to run though the test would
still fail. The test was navigating the iframe in the top frame's load event handler
and expecting a new HistoryItem to get created for the navigation. However, we had
non-standard logic in `NavigationScheduler::mustLockBackForwardList()` that would
lock the back/forward list if any of the iframe's ancestors is still loading. Since
this behavior doesn't match the HTML specification [1][2] or Blink, remove the logic
in mustLockBackForwardList().

The change above caused a different code path to get taken for
navigation-api/navigation-history-entry/entries-after-srcdoc-navigation.html, which
caused it to fail because it would fail to create a new HistoryItem for the srcdoc
navigation. After some investigation, I found that both the back/forward list and the
history list were unlocked. However. we would prevent the creation of a HistoryItem
for a srcdoc navigation because the substitute data created for the navigation had the
`SubstituteData::SessionHistoryVisibility::Hidden` parameter. As per the tests, we
expect srcdoc navigations to be visible in the History list so I am setting this
parameter to "Visible" instead. This got the test passing.

[1] https://html.spec.whatwg.org/#navigate-an-iframe-or-frame (step 2)
[2] https://html.spec.whatwg.org/#completely-finish-loading

* LayoutTests/contentfiltering/block-after-add-data-expected.txt:
* LayoutTests/contentfiltering/block-after-add-data-then-allow-unblock-expected.txt:
* LayoutTests/contentfiltering/block-after-add-data-then-deny-unblock-expected.txt:
* LayoutTests/contentfiltering/block-after-finished-adding-data-expected.txt:
* LayoutTests/contentfiltering/block-after-finished-adding-data-then-allow-unblock-expected.txt:
* LayoutTests/contentfiltering/block-after-finished-adding-data-then-deny-unblock-expected.txt:
* LayoutTests/contentfiltering/block-after-response-expected.txt:
* LayoutTests/contentfiltering/block-after-response-then-allow-unblock-expected.txt:
* LayoutTests/contentfiltering/block-after-response-then-deny-unblock-expected.txt:
* LayoutTests/contentfiltering/block-after-will-send-request-expected.txt:
* LayoutTests/contentfiltering/block-after-will-send-request-then-allow-unblock-expected.txt:
* LayoutTests/contentfiltering/block-after-will-send-request-then-deny-unblock-expected.txt:
* LayoutTests/http/tests/navigation/javascriptlink-subframeload-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/history-traversal/srcdoc/consecutive-srcdoc-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/history-traversal/srcdoc/srcdoc-history-entries-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-history-entry/entries-after-blob-navigation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-history-entry/entries-after-blob-navigation.html:
* LayoutTests/platform/gtk/http/tests/navigation/javascriptlink-frames-expected.txt:
* LayoutTests/platform/mac-ventura/http/tests/navigation/javascriptlink-frames-expected.txt:
* LayoutTests/platform/mac-wk1/contentfiltering/block-after-will-send-request-expected.txt: Copied from LayoutTests/contentfiltering/block-after-will-send-request-expected.txt.
* LayoutTests/platform/mac-wk1/contentfiltering/block-after-will-send-request-then-allow-unblock-expected.txt: Copied from LayoutTests/contentfiltering/block-after-will-send-request-then-allow-unblock-expected.txt.
* LayoutTests/platform/mac-wk1/contentfiltering/block-after-will-send-request-then-deny-unblock-expected.txt: Copied from LayoutTests/contentfiltering/block-after-will-send-request-then-deny-unblock-expected.txt.
* LayoutTests/platform/mac/http/tests/navigation/javascriptlink-frames-expected.txt:
* LayoutTests/platform/wpe/http/tests/navigation/javascriptlink-frames-expected.txt:
* Source/WebCore/html/HTMLFrameElementBase.cpp:
(WebCore::HTMLFrameElementBase::attributeChanged):
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::defaultSubstituteDataForURL):
* Source/WebCore/loader/NavigationScheduler.cpp:
(WebCore::NavigationScheduler::mustLockBackForwardList):

Canonical link: https://commits.webkit.org/287549@main
